### PR TITLE
Add initial Clifford simulator

### PR DESCRIPTION
**Pull request by vtomole**
Author of https://github.com/quantumlib/Cirq/pull/1927 has not been active so I will address the comments in that pull request so we can get this in.

Fixes: https://github.com/quantumlib/Cirq/issues/965

### CODE REVIEW

File: cirq/sim/clifford/ch_form.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union
+import numpy as np
+
+import cirq
+
+
+class CH_Form():
```
Comment (Strilanc) [inline cirq/sim/clifford/ch_form.py]: Rename to `StabilizerStateChForm`

File: cirq/sim/clifford/ch_form.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    r"""
```
Comment (Strilanc) [inline cirq/sim/clifford/ch_form.py]: Nit: summary statement goes on first line.

File: cirq/sim/clifford/ch_form.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    A representation of stabilizer states using the CH form,
+
+        $|\psi> = \omega U_C U_H |s>$
+
+    This representation keeps track of overall phase.
+
+    See Bravyi et al, arXiv:1808.00128 for details.
```
Comment (Strilanc) [inline cirq/sim/clifford/ch_form.py]: Nit: References section

File: cirq/sim/clifford/ch_form.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    """
```
Comment (Strilanc) [inline cirq/sim/clifford/ch_form.py]: Document arguments

File: cirq/sim/clifford/ch_form.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+    def __init__(self, num_qubits, initial_state=0):
```
Comment (Strilanc) [inline cirq/sim/clifford/ch_form.py]: type annotations

File: cirq/sim/clifford/ch_form.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+            while s > 0:
+                yield s & 1
+                s >>= 1
+
+        # Apply X for every non-zero element of initial_state
+        for (i, val) in enumerate(bits(initial_state)):
+            if val:
+                self._X(self.n - i - 1)
+
+    def copy(self):
```
Comment (Strilanc) [inline cirq/sim/clifford/ch_form.py]: types

File: cirq/sim/clifford/ch_form.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+        copy.gamma = self.gamma.copy()
+        copy.v = self.v.copy()
+        copy.s = self.s.copy()
+        copy.omega = self.omega
+
+        return copy
+
+    def __repr__(self):
+        """Return the wavefunction representation of the state."""
+        return cirq.dirac_notation(self.to_state_vector())
```
Comment (Strilanc) [inline cirq/sim/clifford/ch_form.py]: This is not a valid repr. It won't eval to the state. It's more of a `__str__`

File: cirq/sim/clifford/ch_form.py [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+    def _repr_full(self):
+        """Return the CH form representation of the state. """
+        string = ""
+
+        string += "omega: {:.2f}\n".format(self.omega)
```
Comment (Strilanc) [inline cirq/sim/clifford/ch_form.py]: This is much closer to the `__repr__`. `cirq._compact.proper_repr` can handle numpy arrays.

File: cirq/sim/clifford/ch_form.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+        string += "G:\n{}\n".format(np.array(self.G, dtype=int))
+        string += "F:\n{}\n".format(np.array(self.F, dtype=int))
+        string += "M:\n{}\n".format(np.array(self.M, dtype=int))
+        string += "gamma: {}\n".format(np.array(self.gamma, dtype=int))
+        string += "v: {}\n".format(np.array(self.v, dtype=int))
+        string += "s: {}\n".format(np.array(self.s, dtype=int))
+
+        return string
+
+    def inner_product_of_state_and_x(self, x) -> Union[float, complex]:
```
Comment (Strilanc) [inline cirq/sim/clifford/ch_form.py]: input type annotation

File: cirq/sim/clifford/ch_form.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+
+        u = np.zeros(self.n, dtype=bool)
+        for p in range(self.n):
+            if y[p]:
+                u ^= self.F[p, :]
+                mu += 2 * (sum(self.M[p, :] & u) % 2)
+        return self.omega * 2**(-sum(self.v) / 2) * 1j**mu * (
+            -1)**sum(self.v & u & self.s) * np.all(self.v | (u == self.s))
+
+    def wave_function(self):
```
Comment (Strilanc) [inline cirq/sim/clifford/ch_form.py]: types

File: cirq/sim/clifford/clifford_simulator.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+class CliffordSimulator(simulator.SimulatesSamples,
+                        simulator.SimulatesIntermediateState):
+    """An efficient simulator for Clifford circuits."""
+
+    def __init__(self):
+        self.init = True
+
+    def _base_iterator(self, circuit: circuits.Circuit,
+                       qubit_order: ops.QubitOrderOrList,
+                       initial_state: int) -> Iterator:
```
Comment (Strilanc) [inline cirq/sim/clifford/clifford_simulator.py]: Add item type to Iterator

File: cirq/sim/clifford/clifford_simulator.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+
+    Gates and measurements are applied to each representation in O(n^2) time.
+    """
+
+    def __init__(self, qubit_map, initial_state=0):
+        self.qubit_map = qubit_map
+        self.n = len(qubit_map)
+
+        self.tableau = clifford_tableau.CliffordTableau(self.n, initial_state)
+        self.CH_form = ch_form.CH_Form(self.n, initial_state)
```
Comment (Strilanc) [inline cirq/sim/clifford/clifford_simulator.py]: `ch_form`

File: cirq/sim/clifford/ch_form.py (hunk 1 above)
Comment (vtomole) [reply cirq/sim/clifford/ch_form.py]: Done.

File: cirq/sim/clifford/ch_form.py (hunk 2 above)
Comment (vtomole) [reply cirq/sim/clifford/ch_form.py]: Done.

File: cirq/sim/clifford/ch_form.py (hunk 3 above)
Comment (vtomole) [reply cirq/sim/clifford/ch_form.py]: Done.

File: cirq/sim/clifford/ch_form.py (hunk 4 above)
Comment (vtomole) [reply cirq/sim/clifford/ch_form.py]: Done.

File: cirq/sim/clifford/ch_form.py (hunk 5 above)
Comment (vtomole) [reply cirq/sim/clifford/ch_form.py]: Done.

File: cirq/sim/clifford/ch_form.py (hunk 9 above)
Comment (vtomole) [reply cirq/sim/clifford/ch_form.py]: Done.

File: cirq/sim/clifford/ch_form.py (hunk 10 above)
Comment (vtomole) [reply cirq/sim/clifford/ch_form.py]: Done.

File: cirq/sim/clifford/clifford_simulator.py (hunk 12 above)
Comment (vtomole) [reply cirq/sim/clifford/clifford_simulator.py]: Done.

File: cirq/sim/clifford/clifford_simulator.py (hunk 11 above)
Comment (vtomole) [reply cirq/sim/clifford/clifford_simulator.py]: Done.

File: cirq/sim/clifford/ch_form.py (hunk 6 above)
Comment (vtomole) [reply cirq/sim/clifford/ch_form.py]: Did

` def copy(self: 'TSelf') -> 'TSelf':`

**Comment by vtomole** [comment]
@Strilanc I think this is ready to go in as `v0.1`. I'll create an issue to address the improvements we need to make once it's in.

**Pull request merged**